File: go/doc_db_client/doc_db_client_test.go
package doc_db_client

import (
	"context"
	"github.com/muchq/moonbase/protos/doc_db"
	"github.com/stretchr/testify/assert"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"testing"
)

func TestDocDbClient_InsertDocRpcSuccess(t *testing.T) {
	// Arrange
	docEgg := DocEgg{
		Bytes: []byte{1, 2, 3},
		Tags:  make(map[string]string),
	}

	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.InsertDoc("test", docEgg)

	// Assert
	assert.NoError(t, err, "call should have succeeded")
	assert.Equal(t, "123", res.Id)
	assert.Equal(t, "foo", res.Version)
}

func TestDocDbClient_InsertDocRpcFailure(t *testing.T) {
	// Arrange
	docEgg := DocEgg{
		Bytes: []byte{1, 2, 3},
		Tags:  make(map[string]string),
	}

	client := NewDocDbClient(&BrokenStub{}, "foo")

	// Act
	res, err := client.InsertDoc("test", docEgg)

	// Assert
	assert.EqualError(t, err, "rpc error: code = Canceled desc = request canceled")
	assert.Equal(t, res, DocIdAndVersion{}, "res should be zero value")
}

func TestDocDbClient_InsertDocClientValidatesCollection(t *testing.T) {
	// Arrange
	docEgg := DocEgg{
		Bytes: []byte{1, 2, 3},
		Tags:  make(map[string]string),
	}

	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.InsertDoc("", docEgg)

	// Assert
	assert.EqualError(t, err, "collection cannot be empty")
	assert.Equal(t, res, DocIdAndVersion{}, "res should be zero value")
}

func TestDocDbClient_InsertDocClientValidatesBytes(t *testing.T) {
	// Arrange
	docEgg := DocEgg{
		Bytes: []byte{},
		Tags:  make(map[string]string),
	}

	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.InsertDoc("test", docEgg)

	// Assert
	assert.EqualError(t, err, "bytes cannot be empty")
	assert.Equal(t, res, DocIdAndVersion{}, "res should be zero value")
}

func TestDocDbClient_UpdateDocRpcSuccess(t *testing.T) {
	// Arrange
	idAndVersion := DocIdAndVersion{
		Id:      "123",
		Version: "foo",
	}

	docEgg := DocEgg{
		Bytes: []byte{1, 2, 3},
		Tags:  make(map[string]string),
	}

	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.UpdateDoc("test", idAndVersion, docEgg)

	// Assert
	assert.NoError(t, err, "call should have succeeded")
	assert.Equal(t, "123", res.Id)
	assert.Equal(t, "foo_new", res.Version)
}

func TestDocDbClient_UpdateDocRpcFailure(t *testing.T) {
	// Arrange
	idAndVersion := DocIdAndVersion{
		Id:      "123",
		Version: "foo",
	}

	docEgg := DocEgg{
		Bytes: []byte{1, 2, 3},
		Tags:  make(map[string]string),
	}

	client := NewDocDbClient(&BrokenStub{}, "foo")

	// Act
	res, err := client.UpdateDoc("test", idAndVersion, docEgg)

	// Assert
	assert.EqualError(t, err, "rpc error: code = Canceled desc = request canceled")
	assert.Equal(t, res, DocIdAndVersion{}, "res should be zero value")
}

func TestDocDbClient_UpdateDocClientValidatesCollection(t *testing.T) {
	// Arrange
	idAndVersion := DocIdAndVersion{
		Id:      "123",
		Version: "foo",
	}

	docEgg := DocEgg{
		Bytes: []byte{1, 2, 3},
		Tags:  make(map[string]string),
	}

	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.UpdateDoc("", idAndVersion, docEgg)

	// Assert
	assert.EqualError(t, err, "collection cannot be empty")
	assert.Equal(t, res, DocIdAndVersion{}, "res should be zero value")
}

func TestDocDbClient_UpdateDocClientValidatesId(t *testing.T) {
	// Arrange
	idAndVersion := DocIdAndVersion{
		Id:      "",
		Version: "foo",
	}

	docEgg := DocEgg{
		Bytes: []byte{1, 2, 3},
		Tags:  make(map[string]string),
	}

	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.UpdateDoc("test", idAndVersion, docEgg)

	// Assert
	assert.EqualError(t, err, "id cannot be empty")
	assert.Equal(t, res, DocIdAndVersion{}, "res should be zero value")
}

func TestDocDbClient_UpdateDocClientValidatesVersion(t *testing.T) {
	// Arrange
	idAndVersion := DocIdAndVersion{
		Id:      "123",
		Version: "",
	}

	docEgg := DocEgg{
		Bytes: []byte{1, 2, 3},
		Tags:  make(map[string]string),
	}

	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.UpdateDoc("test", idAndVersion, docEgg)

	// Assert
	assert.EqualError(t, err, "version cannot be empty")
	assert.Equal(t, res, DocIdAndVersion{}, "res should be zero value")
}

func TestDocDbClient_UpdateDocClientValidatesBytes(t *testing.T) {
	// Arrange
	idAndVersion := DocIdAndVersion{
		Id:      "123",
		Version: "foo",
	}

	docEgg := DocEgg{
		Bytes: []byte{},
		Tags:  make(map[string]string),
	}

	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.UpdateDoc("test", idAndVersion, docEgg)

	// Assert
	assert.EqualError(t, err, "bytes cannot be empty")
	assert.Equal(t, res, DocIdAndVersion{}, "res should be zero value")
}

func TestDocDbClient_FindDocByIdRpcSuccess(t *testing.T) {
	// Arrange
	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.FindDocById("test", "boing")

	// Assert
	assert.NoError(t, err, "call should have succeeded")
	assert.Equal(t, "boing", res.Id)
	assert.Equal(t, "foo", res.Version)
	assert.Equal(t, []byte{1, 2, 3}, res.Bytes)
	assert.Equal(t, make(map[string]string), res.Tags)
}

func TestDocDbClient_FindDocByIdRpcFailure(t *testing.T) {
	// Arrange
	client := NewDocDbClient(&BrokenStub{}, "foo")

	// Act
	res, err := client.FindDocById("test", "boing")

	// Assert
	assert.EqualError(t, err, "rpc error: code = Canceled desc = request canceled")
	assert.Equal(t, res, Doc{}, "res should be zero value")
}

func TestDocDbClient_FindDocByIdClientValidatesCollection(t *testing.T) {
	// Arrange
	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.FindDocById("", "boing")

	// Assert
	assert.EqualError(t, err, "collection cannot be empty")
	assert.Equal(t, res, Doc{}, "res should be zero value")
}

func TestDocDbClient_FindDocByIdClientValidatesId(t *testing.T) {
	// Arrange
	client := NewDocDbClient(&WorkingStub{}, "foo")

	// Act
	res, err := client.FindDocById("test", "")

	// Assert
	assert.EqualError(t, err, "id cannot be empty")
	assert.Equal(t, res, Doc{}, "res should be zero value")
}

func TestDocDbClient_FindDocByTagsRpcSuccess(t *testing.T) {
	// Arrange
	client := NewDocDbClient(&WorkingStub{}, "foo")
	tags := map[string]string{"foo": "bar"}

	// Act
	res, err := client.FindDocByTags("test", tags)

	// Assert
	assert.NoError(t, err, "call should have succeeded")
	assert.Equal(t, "123", res.Id)
	assert.Equal(t, "foo", res.Version)
	assert.Equal(t, []byte{1, 2, 3}, res.Bytes)
	assert.Equal(t, map[string]string{"foo": "bar"}, res.Tags)
}

func TestDocDbClient_FindDocByTagsRpcFailure(t *testing.T) {
	// Arrange
	client := NewDocDbClient(&BrokenStub{}, "foo")
	tags := map[string]string{"foo": "bar"}

	// Act
	res, err := client.FindDocByTags("test", tags)

	// Assert
	assert.EqualError(t, err, "rpc error: code = Canceled desc = request canceled")
	assert.Equal(t, res, Doc{}, "res should be zero value")
}

func TestDocDbClient_FindDocByTagsClientValidatesCollection(t *testing.T) {
	// Arrange
	client := NewDocDbClient(&WorkingStub{}, "foo")
	tags := map[string]string{"foo": "bar"}

	// Act
	res, err := client.FindDocByTags("", tags)

	// Assert
	assert.EqualError(t, err, "collection cannot be empty")
	assert.Equal(t, res, Doc{}, "res should be zero value")
}

func TestDocDbClient_FindDocByTagsClientValidatesTags(t *testing.T) {
	// Arrange
	client := NewDocDbClient(&WorkingStub{}, "foo")
	tags := make(map[string]string)

	// Act
	res, err := client.FindDocByTags("test", tags)

	// Assert
	assert.EqualError(t, err, "tags cannot be empty")
	assert.Equal(t, res, Doc{}, "res should be zero value")
}

type WorkingStub struct {
}

func (*WorkingStub) InsertDoc(context.Context, *doc_db.InsertDocRequest, ...grpc.CallOption) (*doc_db.InsertDocResponse, error) {
	response := &doc_db.InsertDocResponse{
		Id:      "123",
		Version: "foo",
	}

	return response, nil
}

func (*WorkingStub) UpdateDoc(_ context.Context, req *doc_db.UpdateDocRequest, _ ...grpc.CallOption) (*doc_db.UpdateDocResponse, error) {
	response := &doc_db.UpdateDocResponse{
		Id:      req.Id,
		Version: req.Version + "_new",
	}

	return response, nil
}

func (*WorkingStub) FindDocById(_ context.Context, req *doc_db.FindDocByIdRequest, _ ...grpc.CallOption) (*doc_db.FindDocByIdResponse, error) {
	response := &doc_db.FindDocByIdResponse{
		Doc: &doc_db.Document{
			Id:      req.Id,
			Version: "foo",
			Bytes:   []byte{1, 2, 3},
			Tags:    make(map[string]string),
		},
	}

	return response, nil
}

func (*WorkingStub) FindDoc(_ context.Context, req *doc_db.FindDocRequest, _ ...grpc.CallOption) (*doc_db.FindDocResponse, error) {
	response := &doc_db.FindDocResponse{
		Doc: &doc_db.Document{
			Id:      "123",
			Version: "foo",
			Bytes:   []byte{1, 2, 3},
			Tags:    req.Tags,
		},
	}

	return response, nil
}

type BrokenStub struct {
}

func (*BrokenStub) InsertDoc(context.Context, *doc_db.InsertDocRequest, ...grpc.CallOption) (*doc_db.InsertDocResponse, error) {
	return nil, status.Error(codes.Canceled, "request canceled")
}

func (*BrokenStub) UpdateDoc(context.Context, *doc_db.UpdateDocRequest, ...grpc.CallOption) (*doc_db.UpdateDocResponse, error) {
	return nil, status.Error(codes.Canceled, "request canceled")
}

func (*BrokenStub) FindDocById(context.Context, *doc_db.FindDocByIdRequest, ...grpc.CallOption) (*doc_db.FindDocByIdResponse, error) {
	return nil, status.Error(codes.Canceled, "request canceled")
}

func (*BrokenStub) FindDoc(context.Context, *doc_db.FindDocRequest, ...grpc.CallOption) (*doc_db.FindDocResponse, error) {
	return nil, status.Error(codes.Canceled, "request canceled")
}

File: cpp/doc_db_client/doc_db_client_test.cc
#include "doc_db_client.h"

#include "protos/doc_db/doc_db_mock.grpc.pb.h"

using namespace doc_db;
using ::testing::_;
using ::testing::DoAll;
using ::testing::Return;
using ::testing::SetArgPointee;

DocEgg MakeDocEgg(string bytes, unordered_map<string, string> tags) {
  DocEgg input_doc_egg;
  input_doc_egg.bytes = bytes;
  input_doc_egg.tags = tags;
  return input_doc_egg;
}

DocIdAndVersion MakeInputIds(string id, string version) {
  DocIdAndVersion input_ids;
  input_ids.id = id;
  input_ids.version = version;
  return input_ids;
}

TEST(DocDbClient, InsertDocRpcSuccess) {
  // Arrange
  InsertDocResponse resp;
  resp.set_id("foo");
  resp.set_version("123");

  auto stub = std::make_shared<MockDocDbStub>();
  ON_CALL(*stub, InsertDoc(_, _, _))
      .WillByDefault(DoAll(SetArgPointee<2>(resp), Return(grpc::Status::OK)));

  DocDbClient client(stub, "test");
  DocEgg input_doc_egg = MakeDocEgg("cool bytes", {});

  // Act
  auto status = client.InsertDoc("foo_col", input_doc_egg);

  // Assert
  EXPECT_TRUE(status.ok());
  EXPECT_EQ(status->id, "foo");
  EXPECT_EQ(status->version, "123");
}

TEST(DocDbClient, InsertDocRpcFailure) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  ON_CALL(*stub, InsertDoc(_, _, _)).WillByDefault(Return(grpc::Status::CANCELLED));
  DocDbClient client(stub, "test");
  DocEgg input_doc_egg = MakeDocEgg("cool bytes", {});

  // Act
  auto status = client.InsertDoc("foo_col", input_doc_egg);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::CANCELLED));
}

TEST(DocDbClient, InsertDocClientValidatesCollection) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  DocDbClient client(stub, "test");
  DocEgg input_doc_egg = MakeDocEgg("cool bytes", {});

  // Act
  auto status = client.InsertDoc("", input_doc_egg);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::INVALID_ARGUMENT));
}

TEST(DocDbClient, InsertDocClientValidatesBytes) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  DocDbClient client(stub, "test");
  DocEgg input_doc_egg = MakeDocEgg("", {});

  // Act
  auto status = client.InsertDoc("foo_col", input_doc_egg);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::INVALID_ARGUMENT));
}

TEST(DocDbClient, UpdateDocRpcSuccess) {
  // Arrange
  UpdateDocResponse resp;
  resp.set_id("foo");
  resp.set_version("123");

  auto stub = std::make_shared<MockDocDbStub>();
  ON_CALL(*stub, UpdateDoc(_, _, _))
      .WillByDefault(DoAll(SetArgPointee<2>(resp), Return(grpc::Status::OK)));

  DocDbClient client(stub, "test");
  DocEgg input_doc_egg = MakeDocEgg("cool bytes", {});
  DocIdAndVersion input_id;
  input_id.id = "foo";
  input_id.version = "001";

  // Act
  auto status = client.UpdateDoc("foo_col", input_id, input_doc_egg);

  // Assert
  EXPECT_TRUE(status.ok());
  EXPECT_EQ(status->id, "foo");
  EXPECT_EQ(status->version, "123");
}

TEST(DocDbClient, UpdateDocRpcFailure) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  ON_CALL(*stub, UpdateDoc(_, _, _)).WillByDefault(Return(grpc::Status::CANCELLED));
  DocDbClient client(stub, "test");
  DocEgg input_doc_egg = MakeDocEgg("cool bytes", {});
  DocIdAndVersion input_ids = MakeInputIds("foo", "001");

  // Act
  auto status = client.UpdateDoc("foo_col", input_ids, input_doc_egg);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::CANCELLED));
}

TEST(DocDbClient, UpdateDocClientValidatesCollection) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  DocDbClient client(stub, "test");
  DocEgg input_doc_egg = MakeDocEgg("cool bytes", {});
  DocIdAndVersion input_ids = MakeInputIds("foo", "001");

  // Act
  auto status = client.UpdateDoc("", input_ids, input_doc_egg);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::INVALID_ARGUMENT));
}

TEST(DocDbClient, UpdateDocClientValidatesId) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  DocDbClient client(stub, "test");
  DocEgg input_doc_egg = MakeDocEgg("cool bytes", {});
  DocIdAndVersion input_ids = MakeInputIds("", "001");

  // Act
  auto status = client.UpdateDoc("foo_col", input_ids, input_doc_egg);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::INVALID_ARGUMENT));
}

TEST(DocDbClient, UpdateDocClientValidatesVersion) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  DocDbClient client(stub, "test");
  DocEgg input_doc_egg = MakeDocEgg("cool bytes", {});
  DocIdAndVersion input_ids = MakeInputIds("foo", "");

  // Act
  auto status = client.UpdateDoc("foo_col", input_ids, input_doc_egg);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::INVALID_ARGUMENT));
}

TEST(DocDbClient, UpdateDocClientValidatesBytes) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  DocDbClient client(stub, "test");
  DocEgg input_doc_egg = MakeDocEgg("", {});
  DocIdAndVersion input_ids = MakeInputIds("foo", "001");

  // Act
  auto status = client.UpdateDoc("foo_col", input_ids, input_doc_egg);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::INVALID_ARGUMENT));
}

TEST(DocDbClient, FindDocByIdRpcSuccess) {
  // Arrange
  FindDocByIdResponse resp;
  auto doc = resp.mutable_doc();
  doc->set_id("foo");
  doc->set_version("123");
  doc->set_bytes("neat bytes");

  auto stub = std::make_shared<MockDocDbStub>();
  ON_CALL(*stub, FindDocById(_, _, _))
      .WillByDefault(DoAll(SetArgPointee<2>(resp), Return(grpc::Status::OK)));

  DocDbClient client(stub, "test");

  // Act
  auto status = client.FindDocById("foo_col", "foo");

  // Assert
  EXPECT_TRUE(status.ok());
  EXPECT_EQ(status->id, "foo");
  EXPECT_EQ(status->version, "123");
  EXPECT_EQ(status->bytes, "neat bytes");
  EXPECT_TRUE(status->tags.empty());
}

TEST(DocDbClient, FindDocByIdRpcFailure) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  ON_CALL(*stub, FindDocById(_, _, _)).WillByDefault(Return(grpc::Status::CANCELLED));
  DocDbClient client(stub, "test");

  // Act
  auto status = client.FindDocById("foo_col", "foo");

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::CANCELLED));
}

TEST(DocDbClient, FindDocByIdClientValidatesCollection) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  DocDbClient client(stub, "test");

  // Act
  auto status = client.FindDocById("", "foo");

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::INVALID_ARGUMENT));
}

TEST(DocDbClient, FindDocByIdClientValidatesId) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  DocDbClient client(stub, "test");

  // Act
  auto status = client.FindDocById("foo_col", "");

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::INVALID_ARGUMENT));
}

TEST(DocDbClient, FindDocByTagsRpcSuccess) {
  // Arrange
  FindDocResponse resp;
  auto doc = resp.mutable_doc();
  doc->set_id("foo");
  doc->set_version("123");
  doc->set_bytes("neat bytes");
  auto& tags = *doc->mutable_tags();
  tags["player_1"] = "Tippy";

  auto stub = std::make_shared<MockDocDbStub>();
  ON_CALL(*stub, FindDoc(_, _, _))
      .WillByDefault(DoAll(SetArgPointee<2>(resp), Return(grpc::Status::OK)));

  DocDbClient client(stub, "test");
  unordered_map<string, string> input_tags;
  input_tags["player_1"] = "Tippy";

  // Act
  auto status = client.FindDocByTags("foo_col", input_tags);

  // Assert
  EXPECT_TRUE(status.ok());
  EXPECT_EQ(status->id, "foo");
  EXPECT_EQ(status->version, "123");
  EXPECT_EQ(status->bytes, "neat bytes");
  EXPECT_EQ(status->tags["player_1"], "Tippy");
}

TEST(DocDbClient, FindDocByTagsRpcFailure) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  ON_CALL(*stub, FindDoc(_, _, _)).WillByDefault(Return(grpc::Status::CANCELLED));

  DocDbClient client(stub, "test");
  unordered_map<string, string> input_tags;
  input_tags["player_1"] = "Tippy";

  // Act
  auto status = client.FindDocByTags("foo_col", input_tags);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::CANCELLED));
}

TEST(DocDbClient, FindDocByTagsClientValidatesCollection) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  DocDbClient client(stub, "test");
  unordered_map<string, string> input_tags;
  input_tags["player_1"] = "Tippy";

  // Act
  auto status = client.FindDocByTags("", input_tags);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::INVALID_ARGUMENT));
}

TEST(DocDbClient, FindDocByTagsClientValidatesTags) {
  // Arrange
  auto stub = std::make_shared<MockDocDbStub>();
  DocDbClient client(stub, "test");
  unordered_map<string, string> input_tags;

  // Act
  auto status = client.FindDocByTags("foo_col", input_tags);

  // Assert
  EXPECT_FALSE(status.ok());
  EXPECT_EQ(status.status().code(), absl::StatusCode(grpc::StatusCode::INVALID_ARGUMENT));
}
